test/e2e/clusterctl_upgrade.go

```diff
@@ -575,16 +584,18 @@ func waitForClusterDeletedV1alpha4(ctx context.Context, input waitForClusterDele
 	}, intervals...).Should(BeTrue())
 }
 
-func machinesMatch(oldMachineList *clusterv1alpha3.MachineList, newMachineList *clusterv1.MachineList) bool {
-	if len(oldMachineList.Items) != len(newMachineList.Items) {
+func matchObjects(oldObjects []client.Object, newObjects []client.Object) bool {
```

Comment:
could you clarify the rationale behind going from machinesMatch to matchObjects? which now requires the additional conversion loop

Reply:
Just an idea.

WDYT about working with unstructuredList? Then we don't need the for loops anymore as we would just always work with the UnstructuredList ~:
```go
			postUpgradeMachineList := new(unstructured.UnstructuredList)
			postUpgradeMachineList.SetAPIVersion(clusterv1.GroupVersion.String())
			postUpgradeMachineList.SetKind("MachineList")
                        ...
```
(similar in other places with the corresponding apiVersion of course)

Reply:
Good idea!